src/createScope.tsx

```diff
@@ -18,59 +16,62 @@ export const ScopeProvider = ({
+    const atomSet = new Set(atoms);
+    const createScopedAtom = <T extends AnyWritableAtom>(
+      anAtom: T,
+      delegate: boolean,
+    ): T => {
+      const getAtom = <A extends AnyAtom>(
+        thisArg: AnyAtom,
+        orig: AnyAtom,
+        target: A,
+      ): A => {
+        if (target === thisArg) {
+          return delegate ? getParentScopedAtom(orig as A) : target;
+        }
+        return getScopedAtom(target);
+      };
+      const scopedAtom: typeof anAtom = {
+        ...anAtom,
+        ...('read' in anAtom && {
+          read(get, opts) {
+            return anAtom.read.call(
+              this,
+              (a) => get(getAtom(this, anAtom, a)),
+              opts,
+            );
+          },
+        }),
+        ...('write' in anAtom && {
+          write(get, set, ...args) {
+            return anAtom.write.call(
+              this,
+              (a) => get(getAtom(this, anAtom, a)),
+              (a, ...v) => set(getAtom(this, anAtom, a), ...v),
+              ...args,
+            );
+          },
+        }),
+      };
+      return scopedAtom;
+    };
 
-  const createScopedAtom = <T extends AnyWritableAtom>(
-    anAtom: T,
-    delegate: boolean,
-  ): T => {
-    const getAtom = <A extends AnyAtom>(
-      thisArg: AnyAtom,
-      orig: AnyAtom,
-      target: A,
-    ): A => {
-      if (target === thisArg) {
-        return delegate ? getParentScopedAtom(orig as A) : target;
+    const getScopedAtom: GetScopedAtom = (anAtom) => {
+      let scopedAtom = mapping.get(anAtom);
+      if (!scopedAtom) {
+        scopedAtom = atomSet.has(anAtom)
+          ? createScopedAtom(anAtom as unknown as AnyWritableAtom, false)
+          : createScopedAtom(anAtom as unknown as AnyWritableAtom, true);
+        mapping.set(anAtom, scopedAtom);
       }
-      return getScopedAtom(target);
-    };
-    const scopedAtom: typeof anAtom = {
-      ...anAtom,
-      ...('read' in anAtom && {
-        read(get, opts) {
-          return anAtom.read.call(
-            this,
-            (a) => get(getAtom(this, anAtom, a)),
-            opts,
-          );
-        },
-      }),
-      ...('write' in anAtom && {
-        write(get, set, ...args) {
-          return anAtom.write.call(
-            this,
-            (a) => get(getAtom(this, anAtom, a)),
-            (a, ...v) => set(getAtom(this, anAtom, a), ...v),
-            ...args,
-          );
-        },
-      }),
+      return scopedAtom as typeof anAtom;
     };
-    return scopedAtom;
-  };
 
-  const getScopedAtom: GetScopedAtom = (anAtom) => {
-    let scopedAtom = mapping.get(anAtom);
-    if (!scopedAtom) {
-      scopedAtom = atomSet.has(anAtom)
-        ? createScopedAtom(anAtom as unknown as AnyWritableAtom, false)
-        : createScopedAtom(anAtom as unknown as AnyWritableAtom, true);
-      mapping.set(anAtom, scopedAtom);
-    }
-    return scopedAtom as typeof anAtom;
-  };
+    return getScopedAtom;
+  }, [getParentScopedAtom, ...atoms]);
```

Comment:
I can merge with this change. 
```suggestion
  }, [getParentScopedAtom, atoms]);
```

btw, `atoms` is just an iterable. It doesn't have to be an array.

Reply:
By writting `<ScopeProvider atoms={[a, b, c]}>`, won't the array be renewed during each re-render?

Reply:
You can play with https://codesandbox.io/s/dawn-grass-kzqp3j?file=/src/createScope.js

Reply:
I think you opinion is, use atoms in the dependency array, then remind users memorizing their `atoms` iterable before passing to ScopeProvider in the documentation?

Reply:
Yeah, that's how React suggests in general.
That's why useMemo isn't very useful in this case, but it helps for certain users.
Would you think teaching `memo` is easier?

Reply:
Oh, yesterday I thank about another strategy:
In dev mode, add some useEffect/usePrevious hooks, which warn users if nothing changes but the render body executes again, and instruct them to check the doc.
Then, in the doc instruct users to wrap ScopeProvider with memo to improve performance.

Reply:
That's an interesting idea.
It can be useful for any React libs, not only for our ScopeProvider, right?

Reply:
Huh, you mean build an independent library? There are not so much lines of codes, and as the lib maintainer, you still need to configure the "equality check" yourself, and I am not sure if `process.env.NODE_ENV !== 'production'` is a broadly used flag for non webpack bundled packages.

Reply:
``` typescript
function usePrevious<T>(value: T) {
  const ref = useRef<T | undefined>();
  useEffect(() => {
    ref.current = value;
  }, [value]);
  return ref.current;
}

function useDevModeRenderCheck(msg: string, ...deps: any[], equals?: (a: any, b: any) => boolean) {
  if (process.env.NODE_ENV !== 'production') {
    const prev = usePrevious(deps);
    if (prev !== undefined) {
      if (equals && equals(deps, prev) || !equals && deps.length === prev.length && deps.every((dep, i) => prev[i] === dep) {
        console.warn(msg);
      }
    }
  }
}
```

Reply:
> you mean build an independent library?

I first thought of it too, but it's probably tied to each lib.
What I mean is it's not the jotai-scope specific problem.

Reply:
OK. Apart from that, if this approach is going to be implemented, how can we distinguish unexpected re-render with React.use? By counting to 3?

Reply:
I don't think 3 is safe either. This is just heuristic. Maybe something like 10? Wait, we have Suspense within the tree, so `use` doesn't trigger re-render ScopeProvider, does it?

So, do you plan to remove `useMemo` with the new approach, or keep it?

Reply:
Suspense is fine, so we can warn user to either memo it or add a Suspense.

I'm in favor of the "reminder" approach over `useMemo`. Just keep this PR here until we reach the consensus.

Reply:
Sounds good.